core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -813,7 +813,14 @@ private[spark] class BlockManager(
               false
           }
         } else {
-          memoryStore.putBytes(blockId, size, level.memoryMode, () => bytes)
+          val memoryMode = level.memoryMode
+          memoryStore.putBytes(blockId, size, memoryMode, () => {
+            if (memoryMode == MemoryMode.OFF_HEAP) {
```

Comment:
I did a search. Looks like the buffers passed to `putBytes` (the only caller of private `doPutBytes`) across Spark are all duplicated.